modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java

```diff
@@ -422,65 +422,133 @@ public PathElement[] getCaretShape(int offset, boolean isLeading,
                 x = getMirroringWidth() - x;
             }
             TextLine line = lines[lineIndex];
             TextRun[] runs = line.getRuns();
             RectBounds bounds = line.getBounds();
             TextRun run = null;
-            x -= bounds.getMinX();
             //TODO binary search
             if (text == null || spans == null) {
-                for (int i = 0; i < runs.length; i++) {
-                    run = runs[i];
-                    if (x < run.getWidth()) break;
-                    if (i + 1 < runs.length) {
-                        if (runs[i + 1].isLinebreak()) break;
-                        x -= run.getWidth();
+                // To calculate Text and TextFlow hit info
+                if (isMirrored) {
+                    int runIndex = -1;
+                    for (int i = runs.length - 1; i >=0; i--) {
+                        run = runs[i];
+                        if (x < run.getWidth() && (isTextFlow || (run.getStart() == curRunStart))) {
+                            runIndex = i;
+                            break;
+                        }
+                        if (i - 1 >= 0) {
+                            if (runs[i - 1].isLinebreak()) break;
+                            x -= run.getWidth();
+                        }
+                    }
+                    for (int i = 0; i < runIndex; i++) {
+                        xHitPos -= runs[i].getWidth();
+                    }
+                    xHitPos -= bounds.getMinX();
+                } else {
+                    for (int i = 0; i < runs.length; i++) {
+                        run = runs[i];
+                        if (x < run.getWidth()) break;
+                        if (i + 1 < runs.length) {
+                            if (runs[i + 1].isLinebreak()) break;
+                            x -= run.getWidth();
+                        }
                     }
                 }
             } else {
+                // To calculate hit info of Text embedded in TextFlow
                 for (int i = 0; i < lineIndex; i++) {
                     for (TextRun r: lines[i].runs) {
                         if (r.getTextSpan() != null && r.getStart() >= textRunStart && r.getTextSpan().getText().equals(text)) {
                             textWidthPrevLine += r.getLength();
                         }
                     }
                 }
-                int prevNodeLength = 0;
-                boolean isPrevNodeExcluded = false;
-                for (TextRun r: runs) {
-                    if (!r.getTextSpan().getText().equals(text) || (r.getStart() < textRunStart && r.getTextSpan().getText().equals(text))) {
-                        prevNodeLength += r.getWidth();
-                        continue;
+
+                if (isMirrored) {
+                    for (TextRun r: runs) {
+                        if (r.getStart() != curRunStart && r.getTextSpan().getText().equals(text)) {
+                            isMultiRunText = true;
```

Comment:
minor: we could reduce the scope of `isMultiRunText` by moving its declaration from line 427 to line 490